### DISCUSSION TITLE
Dynamic "assetPrefix" from API

### DISCUSSION THREAD

**Discussion by wuifdesign** [vercel/next.js#38142] [Help]
I have multible CDNs for my production app (based on the domain the app is served from) and want to target the next app to a specific one. 
I get the current CDN URL from an API endpoint i want to call in the "_document" for example.

for example:

- served from a.test.com uses cdn.a.test.com
- served b.test.com uses cdn.b.test.com

is this possible somehow?

**Comment by icyJoseph** [comment]
Using `_document`'s `getInitialProps` sounds like it should work, but also like it'll cause trouble.

I guess one way to do this would be to proxy your requests to a `api/route`, which in turn would check the domain from which the request arrives, and forward the request to the desired CDN. 

In a sense, the solution to this, perhaps should NOT be part of the React tree structure, but part of the supporting server tooling given by Next.js. 

But if we want a React based solution, I could also think of doing preloading, like:

```js
<link rel="preload" href="/api/data" as="fetch" crossorigin="anonymous">
```

Where `api/data` would return the CDN to use for this domain. You can just put this in the `_document` head I believe.

Then you can `fetch('/api/data')`, inside your app, and it'll use the preloaded request. Since it doesn't change per domain, you don't need to refresh or revalidate this data, ever. Though, the process of using this data is still asynchronous. A little pseudo-code follows:

```tsx
const PrefixedImage = ({ src }) => {
  const [prefix, setPrefix] = useState("");

  useEffect(() => {
    const controller = new AbortController();
    const signal = controller.signal;
    // this should use the already preloaded value (perhaps you could store this in a context in _app)
    fetch("/api/data", { signal })
      .then((res) => res.json())
      .then(setPrefix)
      .catch((e) => {
        if (signal.aborted) {
          console.warn("Aborted request");
        } else {
          console.error(e);
        }
      });

    return () => controller.abort();
  }, []);

  if (!prefix) return <ProperFallbackThatPreventsLayoutShift seoFriendly src={src} />;

  return <Asset src={prefix + src} />;
};
```

> useEffect to fetch data is kind of, not recommended anymore, instead, people say we should use data-fetching solutions, like swr, and react-query, but I think for this use case, with the preload link placed, it should kinda work...

I have a few follow ups:

- How do you know in which domain your app is hosted?
- What type of assets are you fetching? Size might be an issue if using `api/routes`
- Are you using `next/image`?

**Reply by wuifdesign** [reply]
hi, thx for your help. 

this may work for images, but not for the js files loaded in the `head`. the only settin affecting this is the `assetPath` (https://nextjs.org/docs/api-reference/next.config.js/cdn-support-with-asset-prefix ). it is only allowed to be a string and i havn't found a way to override it on runtime.

i dont know the domain it is running on. just someone pointing a domain on the next js, then it will do a request to a backend returning the correct cdn. this should than be used for the js files and all other assets.

**Reply by icyJoseph** [reply]
Aha, so it is that type of asset. 

Then I don't know. This is one of the things I fear when it comes to white-label apps. 

What idea did you have with _document?

**Reply by wuifdesign** [reply]
whitelabeling is exactly what i want to do.

i thougt about passing `assetPath` as a prop to the `Head` component or something like this. but haven't found a way.

**Comment by wuifdesign** [comment]
if anyone whats to do this, i found a workaround, it's ugly but it should work:

```tsx
import React, { PropsWithChildren, useContext } from 'react'
import { Html, Head, Main, NextScript } from 'next/document'
import { HtmlContext } from 'next/dist/shared/lib/html-context'

const CustomHtml: React.FC<PropsWithChildren> = ({ children }) => {
  const context = useContext(HtmlContext)
  context.assetPrefix = '.....'
  return (
    <HtmlContext.Provider value={context}>
      <Html>{children}</Html>
    </HtmlContext.Provider>
  )
}

export default function Document() {
  return (
    <CustomHtml>
      <Head />
      <body>
        <Main />
        <NextScript />
      </body>
    </CustomHtml>
  )
}

```

**Reply by justrhysism** [reply]
Absolutely wild that this is even necessary 🙀

**Reply by salazarm** [reply]
Unfortunately this doesn't work with a statically exported app :(

**Reply by salazarm** [reply]
Maybe I can put a placeholder... and have it overwritten..

**Reply by yunsii** [reply]
![image](https://github.com/vercel/next.js/assets/18096089/fdf313e0-f1fe-4c38-a8a7-4afe1689eac9)

Unfortunately, it not work in client.
